django/db/models/constraints.py

```diff
@@ -253,3 +282,57 @@ def deconstruct(self):
         if self.opclasses:
             kwargs['opclasses'] = self.opclasses
         return path, self.expressions, kwargs
+
+    def validate(self, instance, exclude=None, using=DEFAULT_DB_ALIAS):
+        queryset = instance.__class__._default_manager.using(using)
+        if self.fields:
+            lookup_kwargs = {}
+            for field_name in self.fields:
+                if exclude and field_name in exclude:
+                    return
+                field = instance._meta.get_field(field_name)
+                lookup_value = getattr(instance, field.attname)
+                if (lookup_value is None or
+                        (lookup_value == '' and connections[using].features.interprets_empty_strings_as_nulls)):
+                    # A composite constraint containing NULL value cannot caused
+                    # a violation since NULL != NULL in SQL.
+                    return
+                lookup_kwargs[field.name] = lookup_value
+            queryset = queryset.filter(**lookup_kwargs)
+        else:
+            replacement_map = {
+                field.name: Value(getattr(instance, field.attname), field)
+                for field in instance._meta.local_concrete_fields
+                if not exclude or field.name not in exclude
+            }
+            expressions = [
+                Exact(expr, expr.replace_references(replacement_map))
+                for expr in self.expressions
+            ]
+            if exclude:
+                for expression in expressions:
+                    for expr in expression.flatten():
+                        if isinstance(expr, F) and expr.name in exclude:
+                            return
+            queryset = queryset.filter(*expressions)
+        if not instance._state.adding and instance.pk is not None:
+            queryset = queryset.exclude(pk=instance.pk)
```

Comment:
Same comment as prior, about using `_get_pk_val` instead. Disregard this if that comment was irrelevant.

Reply:
It was definitely relevant, let's continue the discussion over there.